distbench_threadpool.cc

```diff
@@ -74,51 +109,151 @@ SimpleThreadpool::~SimpleThreadpool() {
+  }
+  StartWorkerThread();
+}
 
-CThreadpool::~CThreadpool() {
-  thpool_wait(thpool_);
-  thpool_destroy(thpool_);
+void ElasticThreadpool::StartWorkerThread() {
+  {
+    absl::MutexLock m(&work_mutex_);
+    ++threads_launched_;
+    ++active_threads_;
+  }
+  auto elastic_runner = [this]() { TaskRunner(); };
+  RunRegisteredThread("ElasticThreadPool", elastic_runner).detach();
 }
 
-void CThreadpool::AddWork(std::function<void()> function) {
-  // Copy the functor object to the heap, and pass the address of the heap
-  // object to the thread pool:
-  auto* fpointer = new std::function<void()>(function);
-  thpool_add_work(thpool_, Trampoline, fpointer);
+void ElasticThreadpool::AddWork(std::function<void()> function) {
+  auto ok_to_add = [this]() {
+    return (work_queue_.size() + 1) <= max_idle_threads_;
+  };
+  auto ok_conditions = absl::Condition(&ok_to_add);
+
+  absl::MutexLock m(&work_mutex_);
+  work_mutex_.Await(ok_conditions);
+  work_queue_.push(function);
+  ++work_count_;
 }
 
-void CThreadpool::Trampoline(void* heap_object_pointer) {
-  // Execute and delete the heap allocated copy of the functor object:
-  auto f = reinterpret_cast<std::function<void()>*>(heap_object_pointer);
-  (*f)();
-  delete f;
+void ElasticThreadpool::TaskRunner() {
+  bool first_iteration = true;
+  auto work_available = [this]() {
+    return !work_queue_.empty() || shutdown_.HasBeenNotified();
+  };
+  auto working_conditions = absl::Condition(&work_available);
+  std::function<void()> task;
+  bool need_to_grow_threadpool = false;
+  absl::Duration timeout = absl::Milliseconds(10);
+  while (1) {
+    work_mutex_.Lock();
+    if (!first_iteration) {
+      idle_threads_++;
+    }
+    first_iteration = false;
+    if (work_mutex_.AwaitWithTimeout(working_conditions, timeout)) {
+      if (work_queue_.empty()) {
+        --idle_threads_;
+        --active_threads_;
+        work_mutex_.Unlock();
+        return;
+      }
+      // Work is available....
+      need_to_grow_threadpool = work_queue_.size() >= idle_threads_;
+      // need_to_grow_threadpool |= (idle_threads_ == 1);
+      task = std::move(work_queue_.front());
+      work_queue_.pop();
+      // This thread is no longer idle, but.....
+      --idle_threads_;
+      // If we need to launch a new thread we want to account for that upcoming
+      // thread as soon as we drop the work_mutex_, to avoid spawning too
+      // many threads, so we must begin counting it here...
+      if (need_to_grow_threadpool) {
+        ++idle_threads_;
+      }
+    } else {
+      bool need_to_shrink_threadpool = idle_threads_ > max_idle_threads_;
+      bool done = need_to_shrink_threadpool || shutdown_.HasBeenNotified();
+      if (done) {
+        --idle_threads_;
+        --active_threads_;
+        work_mutex_.Unlock();
+        return;
+      } else {
+        work_mutex_.Unlock();
+        continue;
+      }
+    }
+    work_mutex_.Unlock();
+    if (need_to_grow_threadpool) {
```

Comment:
Why start a workerthread if you need to shrink threadpool?

Reply:
This code needs to be cleaned up, so I've reverte this pull request to a draft, but in short the already existing thread can wake up after a period of inactivity and decide to exit if it is no longer needed.